Weapon unique id consistency

**Describe the bug**
Currently, I am trying to store all dropped and picked up weapon items for a given round. The problem is that unique ids of weapons that already dropped and picked up never matched. As far as I understand, weapon unique ids have been created with a weapon entity in item drop and item pickup events again and again.
**To Reproduce**
Any demo file.
**Expected behavior**
The parser should give the same unique id for the same weapon. For Example, if a player dropped a weapon and then other player picked up this weapon, itemdrop and itempickup events should serve with same weapon entity with the same unique id.

You're reactive !

I didn't manage to try it because i don't know how to import a git branch for golang ?

I'm working with docker i tried to replace 
`RUN go get -u github.com/markus-wa/demoinfocs-golang`
by 
```
RUN git clone --single-branch --branch issue-119-ItemPickup-UniqueID https://github.com/markus-wa/demoinfocs-golang.git github.com/markus-wa/demoinfocs-golang
RUN cd github.com/markus-wa/demoinfocs-golang
# Automatically get all dependencies for this project
RUN go get -u ./...
```
 
But i got dependencies errors :

> Step 9/12 : RUN go get -u ./...
>  ---> Running in 182dfcbb564c
> github.com/dustin/go-heatmap/examples/wikipedia/wikipedia.go:14:2: cannot find package "github.com/dustin/go-humanize" in any of:
> 	/usr/local/go/src/github.com/dustin/go-humanize (from $GOROOT)
> 	/go/src/github.com/dustin/go-humanize (from $GOPATH)
> github.com/llgcode/draw2d/draw2dgl/gc.go:11:2: cannot find package "github.com/go-gl/gl/v2.1/gl" in any of:
> 	/usr/local/go/src/github.com/go-gl/gl/v2.1/gl (from $GOROOT)
> 	/go/src/github.com/go-gl/gl/v2.1/gl (from $GOPATH)
> github.com/llgcode/draw2d/draw2dpdf/fileutil.go:6:8: cannot find package "github.com/jung-kurt/gofpdf" in any of:
> 	/usr/local/go/src/github.com/jung-kurt/gofpdf (from $GOROOT)
> 	/go/src/github.com/jung-kurt/gofpdf (from $GOPATH)
> github.com/llgcode/draw2d/samples/helloworldgl/helloworldgl.go:10:2: cannot find package "github.com/go-gl/glfw/v3.1/glfw" in any of:
> 	/usr/local/go/src/github.com/go-gl/glfw/v3.1/glfw (from $GOROOT)
> 	/go/src/github.com/go-gl/glfw/v3.1/glfw (from $GOPATH)
> github.com/llgcode/draw2d/samples/postscript/postscript.go:9:2: cannot find package "github.com/llgcode/ps" in any of:
> 	/usr/local/go/src/github.com/llgcode/ps (from $GOROOT)
> 	/go/src/github.com/llgcode/ps (from $GOPATH)
> golang.org/x/image/font/sfnt/cmap.go:8:2: cannot find package "golang.org/x/text/encoding/charmap" in any of:
> 	/usr/local/go/src/golang.org/x/text/encoding/charmap (from $GOROOT)
> 	/go/src/golang.org/x/text/encoding/charmap (from $GOPATH)
> golang.org/x/tools/godoc/redirect/redirect.go:22:2: cannot find package "golang.org/x/net/context/ctxhttp" in any of:
> 	/usr/local/go/src/golang.org/x/net/context/ctxhttp (from $GOROOT)
> 	/go/src/golang.org/x/net/context/ctxhttp (from $GOPATH)
> golang.org/x/tools/internal/lsp/source/analysis.go:20:2: cannot find package "golang.org/x/sync/errgroup" in any of:
> 	/usr/local/go/src/golang.org/x/sync/errgroup (from $GOROOT)
> 	/go/src/golang.org/x/sync/errgroup (from $GOPATH)
> golang.org/x/tools/cmd/html2article/conv.go:21:2: cannot find package "golang.org/x/net/html" in any of:
> 	/usr/local/go/src/golang.org/x/net/html (from $GOROOT)
> 	/go/src/golang.org/x/net/html (from $GOPATH)
> golang.org/x/tools/cmd/html2article/conv.go:22:2: cannot find package "golang.org/x/net/html/atom" in any of:
> 	/usr/local/go/src/golang.org/x/net/html/atom (from $GOROOT)
> 	/go/src/golang.org/x/net/html/atom (from $GOPATH)
> golang.org/x/tools/playground/socket/socket.go:37:2: cannot find package "golang.org/x/net/websocket" in any of:
> 	/usr/local/go/src/golang.org/x/net/websocket (from $GOROOT)
> 	/go/src/golang.org/x/net/websocket (from $GOPATH)
> Service 'analyzer' failed to build: The command '/bin/sh -c go get -u ./...' returned a non-zero code: 1

I might have tried this already but got this error :

> go: cannot use path@version syntax in GOPATH mode

But with `GO111MODULE=on` environment variable added it works 👍 

However i don't understand how to access WeaponTraceable() ?

I tried : 
```go
    p.RegisterEventHandler(func(e events.ItemDrop) {

        if(isGameLive) {

            fmt.Printf("Player %s Dropped weapon %s (%d)\n", e.Player.Name, e.Weapon.Weapon.String(), e.WeaponTraceable().UniqueID())
        }
    })

    p.RegisterEventHandler(func(e events.ItemPickup) {

        if(isGameLive) {

            fmt.Printf("Player %s Picked up weapon %s (%d)\n", e.Player.Name, e.Weapon.Weapon.String(), e.WeaponTraceable().UniqueID())
        }
    })
```

Error : 

> app/app.go:54:104: e.WeaponTraceable undefined (type events.ItemDrop has no field or method WeaponTraceable)
> app/app.go:62:106: e.WeaponTraceable undefined (type events.ItemPickup has no field or method WeaponTraceable)
>

Sorry, i tried everything i could without any result ...

Here's my Dockerfile (with old tests disabled) :
```
FROM golang:1.12-alpine

# Update package list
RUN apk update

# Install git
RUN apk add git

# Install vgo to use specific version of packages (See: https://godoc.org/golang.org/x/vgo)
#RUN go get -u golang.org/x/vgo

#COPY src /go/src
WORKDIR /app/

# init module for the project
RUN go mod init demo-analyzer

# TMP TEST BRANCH
#RUN vgo get github.com/markus-wa/demoinfocs-golang@issue-119-ItemPickup-UniqueID
#RUN go get gopkg.in/markus-wa/demoinfocs-golang.issue-119-ItemPickup-UniqueID
RUN go get github.com/markus-wa/demoinfocs-golang@issue-119-ItemPickup-UniqueID

CMD ["go", "run", "/app/app.go"]
``` 

Docker output (it does build) : 

> Building analyzer
> Step 1/8 : FROM golang:1.12-alpine
>  ---> 4e4b7a8b9495
> Step 2/8 : RUN apk update
>  ---> Using cache
>  ---> 63615f87ceef
> Step 3/8 : RUN apk add git
>  ---> Using cache
>  ---> d0436b3670c0
> Step 4/8 : WORKDIR /app/
>  ---> Running in c6159157a115
> Removing intermediate container c6159157a115
>  ---> 23a264e21fb0
> Step 5/8 : RUN go mod init demo-analyzer
>  ---> Running in 1aa51eb71bd0
> go: creating new go.mod: module demo-analyzer
> Removing intermediate container 1aa51eb71bd0
>  ---> a3b550f53091
> Step 6/8 : RUN go get github.com/markus-wa/demoinfocs-golang@issue-119-ItemPickup-UniqueID
>  ---> Running in 10216bc113b2
> go: finding github.com/markus-wa/demoinfocs-golang issue-119-ItemPickup-UniqueID
> go: finding github.com/dustin/go-heatmap v0.0.0-20180603032536-b89dbd73785a
> go: finding github.com/golang/freetype v0.0.0-20170609003504-e2365dfdc4a0
> go: finding github.com/llgcode/draw2d v0.0.0-20180124133339-274031cf2abe
> go: finding github.com/markus-wa/quickhull-go v0.0.0-20190116183559-9fb9702adbda
> go: finding github.com/gogo/protobuf v1.1.1
> go: finding github.com/llgcode/ps v0.0.0-20150911083025-f1443b32eedb
> go: finding github.com/stretchr/objx v0.1.1
> go: finding github.com/markus-wa/godispatch v1.1.0
> go: finding github.com/stretchr/testify v1.2.2
> go: finding golang.org/x/image v0.0.0-20180708004352-c73c2afc3b81
> go: finding github.com/markus-wa/gobitread v0.2.2
> go: finding github.com/pmezard/go-difflib v1.0.0
> go: finding github.com/davecgh/go-spew v1.1.1
> go: finding github.com/golang/geo v0.0.0-20180826223333-635502111454
> go: downloading github.com/markus-wa/demoinfocs-golang v1.2.2-0.20190718222654-0b073fe5c643
> go: extracting github.com/markus-wa/demoinfocs-golang v1.2.2-0.20190718222654-0b073fe5c643
> go: downloading github.com/golang/geo v0.0.0-20180826223333-635502111454
> go: downloading github.com/gogo/protobuf v1.1.1
> go: downloading github.com/markus-wa/godispatch v1.1.0
> go: downloading github.com/markus-wa/quickhull-go v0.0.0-20190116183559-9fb9702adbda
> go: extracting github.com/markus-wa/godispatch v1.1.0
> go: extracting github.com/markus-wa/quickhull-go v0.0.0-20190116183559-9fb9702adbda
> go: downloading github.com/markus-wa/gobitread v0.2.2
> go: extracting github.com/markus-wa/gobitread v0.2.2
> go: extracting github.com/golang/geo v0.0.0-20180826223333-635502111454
> go: extracting github.com/gogo/protobuf v1.1.1
> Removing intermediate container 10216bc113b2
>  ---> be66c6a6b5bd
> Step 7/8 : EXPOSE 80
>  ---> Running in 633aff9670b7
> Removing intermediate container 633aff9670b7
>  ---> 16e22f3a1e5f
> Step 8/8 : CMD ["go", "run", "/app/app.go"]
>  ---> Running in 71966e43d552
> Removing intermediate container 71966e43d552
>  ---> e2240c086fcd
> 
> Successfully built e2240c086fcd


Error on execution : 

> app.go:8:2: cannot find package "github.com/markus-wa/demoinfocs-golang" in any of:
> 	/usr/local/go/src/github.com/markus-wa/demoinfocs-golang (from $GOROOT)
> 	/go/src/github.com/markus-wa/demoinfocs-golang (from $GOPATH)
> app.go:9:2: cannot find package "github.com/markus-wa/demoinfocs-golang/events" in any of:
> 	/usr/local/go/src/github.com/markus-wa/demoinfocs-golang/events (from $GOROOT)
> 	/go/src/github.com/markus-wa/demoinfocs-golang/events (from $GOPATH)

Your package is not stored in `/go/src/` or `/usr/local/go/src/` and there is no `go.mod` & `go.sum` inside my module.

--- a/events/events.go
+++ b/events/events.go
@@ -407,22 +407,42 @@ type RankUpdate struct {
 // ItemEquip signals an item was equipped.
 // This event is not available in all demos.
 type ItemEquip struct {
-	Weapon common.Equipment
-	Player *common.Player
+	Player    *common.Player
+	Weapon    common.Equipment // Value of WeaponPtr, use WeaponPtr instead
+	WeaponPtr *common.Equipment
 }
 
 // ItemPickup signals an item was bought or picked up.
 // This event is not available in all demos.
 type ItemPickup struct {
-	Weapon common.Equipment
 	Player *common.Player
+	Weapon common.Equipment // UniqueID() of this cannot be used to trace the weapon, use WeaponTraceable() instead
+}
+
+// WeaponTraceable attempts to return the original Equipment instance.
+// This can be used to trace the same weapon with UniqueID() when it's being dropped to other players.
+//
+// If the original instance cannot be found, the returned Equipment will still have a different UniqueID().
+func (pu ItemPickup) WeaponTraceable() *common.Equipment {
+	if pu.Player == nil {
+		return &pu.Weapon
+	}
+
+	for _, wep := range pu.Player.RawWeapons {
+		if wep.Weapon == pu.Weapon.Weapon {
+			return wep
+		}
+	}
+
+	return &pu.Weapon
 }
 
 // ItemDrop signals an item was dropped.
 // This event is not available in all demos.
 type ItemDrop struct {
-	Weapon common.Equipment
-	Player *common.Player
+	Player    *common.Player
+	Weapon    common.Equipment // Value of WeaponPtr, use WeaponPtr instead
+	WeaponPtr *common.Equipment
 }
 
 // DataTablesParsed signals that the datatables were parsed.

--- a/events/events_test.go
+++ b/events/events_test.go
@@ -42,6 +42,45 @@ func TestBombEvents(t *testing.T) {
 	}
 }
 
+func TestItemPickup_WeaponTraceable_PlayerNil(t *testing.T) {
+	e := ItemPickup{
+		Weapon: common.Equipment{Weapon: common.EqAK47},
+		Player: nil,
+	}
+
+	assert.Equal(t, e.Weapon, *e.WeaponTraceable())
+}
+
+func TestItemPickup_WeaponTraceable_WeaponFound(t *testing.T) {
+	wep := &common.Equipment{
+		EntityID: 1,
+		Weapon:   common.EqAK47,
+	}
+	e := ItemPickup{
+		Weapon: common.Equipment{Weapon: common.EqAK47},
+		Player: &common.Player{RawWeapons: map[int]*common.Equipment{
+			1: wep,
+		}},
+	}
+
+	assert.Equal(t, wep, e.WeaponTraceable())
+}
+
+func TestItemPickup_WeaponTraceable_WeaponNotFound(t *testing.T) {
+	wep := &common.Equipment{
+		EntityID: 1,
+		Weapon:   common.EqAK47,
+	}
+	e := ItemPickup{
+		Weapon: common.Equipment{Weapon: common.EqKnife},
+		Player: &common.Player{RawWeapons: map[int]*common.Equipment{
+			1: wep,
+		}},
+	}
+
+	assert.Equal(t, e.Weapon, *e.WeaponTraceable())
+}
+
 type demoInfoProvider struct {
 }
 

--- a/game_events.go
+++ b/game_events.go
@@ -481,33 +481,36 @@ func (geh gameEventHandler) bombBegindefuse(desc *msg.CSVCMsg_GameEventListDescr
 func (geh gameEventHandler) itemEquip(desc *msg.CSVCMsg_GameEventListDescriptorT, ge *msg.CSVCMsg_GameEvent) {
 	player, weapon := geh.itemEvent(desc, ge)
 	geh.dispatch(events.ItemEquip{
-		Player: player,
-		Weapon: weapon,
+		Player:    player,
+		Weapon:    *weapon,
+		WeaponPtr: weapon,
 	})
 }
 
 func (geh gameEventHandler) itemPickup(desc *msg.CSVCMsg_GameEventListDescriptorT, ge *msg.CSVCMsg_GameEvent) {
 	player, weapon := geh.itemEvent(desc, ge)
-	geh.dispatch(events.ItemPickup{
+	// Delayed because of #119 - Equipment.UniqueID()
+	geh.parser.delayedEvents = append(geh.parser.delayedEvents, events.ItemPickup{
 		Player: player,
-		Weapon: weapon,
+		Weapon: *weapon,
 	})
 }
 
 func (geh gameEventHandler) itemRemove(desc *msg.CSVCMsg_GameEventListDescriptorT, ge *msg.CSVCMsg_GameEvent) {
 	player, weapon := geh.itemEvent(desc, ge)
 	geh.dispatch(events.ItemDrop{
-		Player: player,
-		Weapon: weapon,
+		Player:    player,
+		Weapon:    *weapon,
+		WeaponPtr: weapon,
 	})
 }
 
-func (geh gameEventHandler) itemEvent(desc *msg.CSVCMsg_GameEventListDescriptorT, ge *msg.CSVCMsg_GameEvent) (*common.Player, common.Equipment) {
+func (geh gameEventHandler) itemEvent(desc *msg.CSVCMsg_GameEventListDescriptorT, ge *msg.CSVCMsg_GameEvent) (*common.Player, *common.Equipment) {
 	data := mapGameEventData(desc, ge)
 	player := geh.playerByUserID32(data["userid"].GetValShort())
 
 	wepType := common.MapEquipment(data["item"].GetValString())
-	weapon := common.NewEquipment(wepType)
+	weapon := getPlayerWeapon(player, wepType)
 
 	return player, weapon
 }

